### PR TITLE
[MRG+1] Add SPoC spatial filtering for continuous target decoding

### PR DESCRIPTION
**Pull request by alexandrebarachant**
Source Power Comodulation (SPoC) allows to identify the composition of
orthogonal spatial filters that maximally correlate with a continuous target.
SPoC can be seen as an extension of the CSP for continuous variables.

We also added a new dataset to illustrates this method in an example consisting in decoding left hand EMG power from MEG data. The dataset is fetched from a fieldtrip tutorial (http://www.fieldtriptoolbox.org/tutorial/coherence)

the output looks like this : 
figure_1-1

### CODE REVIEW

File: mne/datasets/fieldtrip_cmc/__init__.py [hunk 1]
```diff
@@ -0,0 +1,3 @@
+"""fieldtrip CMC Dataset."""
```
Comment (agramfort) [inline mne/datasets/fieldtrip_cmc/__init__.py]: CMC -> Cortico-Muscular Coherence (CMC)

File: mne/decoding/csp.py [hunk 2]
```diff
@@ -652,3 +656,150 @@ def _ajd_pham(X, eps=1e-6, max_iter=15):
...
+        applies a log transform to standardize the features, else the features
+        are z-scored. If transform_into == 'csp_space', then log must be None.
+    transform_into : {'average_power', 'csp_space'}
+        If 'average_power' then self.transform will return the average power of
+        each spatial filter. If 'csp_space' self.transform will return the data
+        in CSP space. Defaults to 'average_power'.
+
+    Attributes
+    ----------
+    ``filters_`` : ndarray, shape(n_components, n_channels)
```
Comment (agramfort) [inline mne/decoding/csp.py]: missing space after shape

File: mne/decoding/csp.py [hunk 3]
```diff
@@ -652,3 +656,150 @@ def _ajd_pham(X, eps=1e-6, max_iter=15):
...
+        If fit, the SPoC spatial filters, else None.
+    ``patterns_`` : ndarray, shape(n_components, n_channels)
```
Comment (agramfort) [inline mne/decoding/csp.py]: missing space after shape

File: mne/decoding/csp.py [hunk 4]
```diff
@@ -652,3 +656,150 @@ def _ajd_pham(X, eps=1e-6, max_iter=15):
...
+        -------
+        self : instance of SPoC
+            Returns the modified instance.
+        """
+        if not isinstance(X, np.ndarray):
+            raise ValueError("X should be of type ndarray (got %s)."
+                             % type(X))
+        self._check_Xy(X, y)
+
+        if len(set(y)) < 2:
```
Comment (agramfort) [inline mne/decoding/csp.py]: use np.unique

File: mne/decoding/csp.py [hunk 5]
```diff
@@ -652,3 +656,150 @@ def _ajd_pham(X, eps=1e-6, max_iter=15):
...
+            raise ValueError("y must have at least two distinct values.")
+
+        # The following code is direclty copied from pyRiemann
+
+        # Normalize target variable
+        target = np.float64(y.copy())
```
Comment (agramfort) [inline mne/decoding/csp.py]: y.copy().astype(np.float64)

Comment (Eric89GXL) [inline mne/decoding/csp.py]: `astype` makes a copy already, so just `y.astype(np.float64)`

File: mne/decoding/csp.py [hunk 6]
```diff
@@ -652,3 +656,150 @@ def _ajd_pham(X, eps=1e-6, max_iter=15):
...
+        target -= target.mean()
+        target /= target.std()
+
+        n_epochs, n_channels = X.shape[:2]
+
+        # Estimate single trial covariance
+        covs = np.zeros((n_epochs, n_channels, n_channels))
```
Comment (agramfort) [inline mne/decoding/csp.py]: np.empty

File: mne/decoding/csp.py [hunk 7]
```diff
@@ -652,3 +656,150 @@ def _ajd_pham(X, eps=1e-6, max_iter=15):
...
+        for ii, epoch in enumerate(X):
+            covs[ii] = _regularized_covariance(epoch, reg=self.reg)
+
+        C = covs.mean(0)
+        Cz = np.mean(covs * target[:, np.newaxis, np.newaxis], axis=0)
+
+        # solve eigenvalue decomposition
+        evals, evecs = linalg.eigh(Cz, C)
+        evals = evals.real  # XXX check
```
Comment (agramfort) [inline mne/decoding/csp.py]: XXX ?

**Comment by alexandrebarachant** [comment]
circle is happy :) 

https://4247-1301584-gh.circle-artifacts.com/0/home/ubuntu/mne-python/doc/_build/html/auto_examples/decoding/plot_decoding_spoc_CMC.html

File: examples/decoding/plot_decoding_spoc_CMC.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+emg = raw.copy().pick_channels(['EMGlft'])
+emg.filter(20., None)
+
+# Filter MEG data to focus on alpha band
+raw.pick_types(meg=True, ref_meg=True, eeg=False, eog=False)
+raw.filter(15., 30., method='iir')
+
+# Build epochs as sliding windows over the continuous raw file
+onsets = raw.first_samp + np.arange(0., raw.n_times, .250 * raw.info['sfreq'])
+events = np.c_[onsets, np.zeros((len(onsets), 2))].astype(int)
```
Comment (agramfort) [inline examples/decoding/plot_decoding_spoc_CMC.py]: events = np.c_[onsets, np.zeros((len(onsets), 2), dtype=int)]

should make it int

**Comment by agramfort** [comment]
besides my nitpick +1 for MRG when CIs are happy

File: doc/manual/decoding.rst [hunk 9]
```diff
@@ -75,6 +75,17 @@ Large entries in the diagonal matrix corresponds to a spatial filter which gives
 
     The winning entry of the Grasp-and-lift EEG competition in Kaggle uses the CSP implementation in MNE. It was featured as a `script of the week`_.
 
+
+Source Power Comodulation (SPoC)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Source Power Comodulation (SPoC) [1] allows to identify the composition of orthogonal spatial filters that maximally correlate with a continuous target.
```
Comment (Eric89GXL) [inline doc/manual/decoding.rst]: [1]_

then later put it in a `references` section. See how it's done e.g. in `doc/manual/datasets_index.rst`

File: examples/decoding/plot_decoding_spoc_CMC.py [hunk 10]
```diff
@@ -0,0 +1,87 @@
+"""
+====================================
+Continuous Target Decoding with SPoC
+====================================
+
+Source Power Comodulation (SPoC) [1] allows to identify the composition of
```
Comment (Eric89GXL) [inline examples/decoding/plot_decoding_spoc_CMC.py]: [1]_

File: examples/decoding/plot_decoding_spoc_CMC.py [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+orthogonal spatial filters that maximally correlate with a continuous target.
+
+SPoC can be seen as an extension of the CSP for continuous variables.
+
+Here, SPoC is applied to decode the (continuous) fluctuation of an
+electromyogram from MEG beta activity [2].
```
Comment (Eric89GXL) [inline examples/decoding/plot_decoding_spoc_CMC.py]: This would be [2]_ but it's a website, not an academic reference, so let's just use a ReST link here instead.

File: mne/datasets/utils.py [hunk 12]
```diff
@@ -271,7 +274,8 @@ def _data_path(path=None, force_update=False, update_path=True, download=True,
                 'tar.gz/%s' % releases['testing'],
         multimodal='https://ndownloader.figshare.com/files/5999598',
         visual_92_categories='https://mne-tools.s3.amazonaws.com/datasets/%s',
-        mtrf="https://superb-dca2.dl.sourceforge.net/project/aespa/%s"
+        mtrf="https://superb-dca2.dl.sourceforge.net/project/aespa/%s",
+        fieldtrip_cmc="ftp://ftp.fieldtriptoolbox.org/pub/fieldtrip/tutorial/%s"
```
Comment (Eric89GXL) [inline mne/datasets/utils.py]: please also modify `circle.yml` to autodetect when to download this dataset. You can see where the conditionals are for our other datasets

File: mne/decoding/csp.py [hunk 13]
```diff
@@ -652,3 +656,150 @@ def _ajd_pham(X, eps=1e-6, max_iter=15):
             break
     D = np.reshape(A, (n_times, -1, n_times)).transpose(1, 0, 2)
     return V, D
+
+
+class SPoC(CSP):
+    """Implementation of the SPoC spatial filtering with Covariance as input.
```
Comment (Eric89GXL) [inline mne/decoding/csp.py]: I don't think you need to say "with covariance as input", it doesn't add much. It would be better to do:
```
"""
Source Power Comodulation (SPoC) spatial filtering.

SPoC [1]_ allows ...
```

File: mne/decoding/csp.py [hunk 14]
```diff
@@ -652,3 +656,150 @@ def _ajd_pham(X, eps=1e-6, max_iter=15):
...
+
+    Source Power Comodulation (SPoC) [1] allows to extract spatial filters and
```
Comment (Eric89GXL) [inline mne/decoding/csp.py]: [1]_

File: mne/decoding/csp.py [hunk 15]
```diff
@@ -652,3 +656,150 @@ def _ajd_pham(X, eps=1e-6, max_iter=15):
...
+    ``std_`` : ndarray, shape (n_components,)
+        If fit, the std squared power for each component.
+
+    See Also
+    --------
+    mne.preprocessing.Xdawn, CSP
+
+    References
+    ----------
+    [1] Dahne, S., Meinecke, F. C., Haufe, S., Hohne, J., Tangermann, M.,
```
Comment (Eric89GXL) [inline mne/decoding/csp.py]: .. [1]

(these are NumPy doc standards FYI)

File: mne/decoding/csp.py [hunk 16]
```diff
@@ -652,3 +656,150 @@ def _ajd_pham(X, eps=1e-6, max_iter=15):
...
+        relevant parameters. NeuroImage, 86, 111-122.
+    """
+
+    def __init__(self, n_components=4, reg=None, log=None,
+                 transform_into='average_power'):
+        """Init of SPoC."""
+        super(SPoC, self).__init__(n_components=n_components, reg=reg, log=log,
+                                   cov_est="epoch",
+                                   transform_into=transform_into)
+        delattr(self, 'cov_est')
```
Comment (Eric89GXL) [inline mne/decoding/csp.py]: Add a comment why (presumably because it gets estimated per-epoch internally?)

File: mne/decoding/csp.py [hunk 17]
```diff
@@ -652,3 +656,150 @@ def _ajd_pham(X, eps=1e-6, max_iter=15):
...
+        evals = evals.real
+        evecs = evecs.real
+        # sort vectors
+        ix = np.argsort(np.abs(evals))[::-1]
+
+        # sort eigenvectors
+        evecs = evecs[:, ix].T
+
+        # spatial patterns
+        self.patterns_ = linalg.pinv(evecs).T  # n_components x n_channels
```
Comment (Eric89GXL) [inline mne/decoding/csp.py]: isn't `np.allclose(linalg.pinv(evecs), evecs.T)` (by orthonormality) and thus `np.allclose(linalg.pinv(evecs).T, evecs)`?

Comment (Eric89GXL) [inline mne/decoding/csp.py]: ... or if this is to take care of some rank deficiency, you could/should operate on `evals` directly to determine the cutoff

File: mne/decoding/tests/test_csp.py [hunk 18]
```diff
@@ -201,3 +201,25 @@ def test_ajd():
...
+    spoc = SPoC(n_components=4)
+    spoc.fit(X, y)
+    Xt = spoc.transform(X)
+    assert_array_equal(Xt.shape, [10, 4])
+    spoc = SPoC(n_components=4, transform_into='csp_space')
+    spoc.fit(X, y)
+    Xt = spoc.transform(X)
+    assert_array_equal(Xt.shape, [10, 4, 20])
+    assert_array_equal(spoc.filters_.shape, [10, 10])
+    assert_array_equal(spoc.patterns_.shape, [10, 10])
```
Comment (Eric89GXL) [inline mne/decoding/tests/test_csp.py]: How hard would it be to simulate some signal that should clearly work with the algorithm, then verify it works? That would be one step better than these smoke tests that just check shape.

File: examples/decoding/plot_decoding_spoc_CMC.py [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+# Filter MEG data to focus on alpha band
+raw.pick_types(meg=True, ref_meg=True, eeg=False, eog=False)
+raw.filter(15., 30., method='iir')
+
+# Build epochs as sliding windows over the continuous raw file
+step = int(.250 * raw.info['sfreq'])
+onsets = raw.first_samp + np.arange(0, raw.n_times, step)
+events = np.c_[onsets, np.zeros((len(onsets), 2), dtype=int)]
+
+# Epoch lenght is 1.5 second
```
Comment (jona-sassenhagen) [inline examples/decoding/plot_decoding_spoc_CMC.py]: length

File: examples/decoding/plot_decoding_spoc_CMC.py [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+emg.filter(20., None)
+
+# Filter MEG data to focus on alpha band
+raw.pick_types(meg=True, ref_meg=True, eeg=False, eog=False)
+raw.filter(15., 30., method='iir')
+
+# Build epochs as sliding windows over the continuous raw file
+step = int(.250 * raw.info['sfreq'])
+onsets = raw.first_samp + np.arange(0, raw.n_times, step)
+events = np.c_[onsets, np.zeros((len(onsets), 2), dtype=int)]
```
Comment (jona-sassenhagen) [inline examples/decoding/plot_decoding_spoc_CMC.py]: `mne.make_fixed_length_events`?

File: mne/datasets/utils.py (hunk 12 above)
Comment (alexandrebarachant) [inline mne/datasets/utils.py]: Can you check circle.yml, i made the modifications

**Comment by agramfort** [comment]
@alexandrebarachant you need to rebase

**Comment by agramfort** [comment]
@robertoostenveld FYI we now fetch a fieldtrip dataset to illustrate this SPoC method.

**Comment by robertoostenveld** [comment]
thanks for the notification.

**Comment by agramfort** [comment]
@robertoostenveld of course. Let us know if we can have a better citation to give credit for sharing this data.

File: mne/decoding/tests/test_csp.py (hunk 18 above)
Comment (jona-sassenhagen) [inline mne/decoding/tests/test_csp.py]: You could just draw random EEG data e, take a random topo M, your y is dot(M, e) + noise, and the correlation between the estimated pattern and M should be high. Right?

**Comment by Eric89GXL** [comment]
circle modifications look correct to me

Just the testing issue remains I think.

**Comment by agramfort** [comment]
@Eric89GXL merge if you're happy

File: mne/decoding/tests/test_csp.py (hunk 18 above)
Comment (alexandrebarachant) [inline mne/decoding/tests/test_csp.py]: We have code for simulation since we were originally working on an example with  simulated data.
If i can make it compact enough for the test, i will push it.

File: mne/decoding/tests/test_csp.py [hunk 21]
```diff
@@ -132,6 +154,22 @@ def test_csp():
...
+    y = np.array([100] * 50 + [1] * 50)
+    X, A = simulate_data(y)
+
+    # fit spoc
+    csp = CSP(n_components=3)
+    csp.fit(X, y)
+
+    # get the patterns and normalize them
+    patterns = csp.patterns_.T
+    patterns = patterns / np.apply_along_axis(np.linalg.norm, 0, patterns)
```
Comment (Eric89GXL) [inline mne/decoding/tests/test_csp.py]: why not `patterns /= np.linalg.norm(patterns, axis=0, keepdims=True)`?

Comment (Eric89GXL) [inline mne/decoding/tests/test_csp.py]: (for full equivalence you might want a `.copy()` above you if adopt this)

File: mne/decoding/tests/test_csp.py [hunk 22]
```diff
@@ -23,6 +23,28 @@
...
+
+    Data are simulated in the statistical source space, where one source is
+    modulated according to a target variable, before being mixed with a
+    random mixing matrix.
+    """
+    rs = np.random.RandomState(random_state)
+
+    # generate a orthogonal mixin matrix
+    mixing_mat = rs.randn(n_channels, n_channels)
+    _, mixing_mat = np.linalg.eigh(np.dot(mixing_mat.T, mixing_mat))
```
Comment (Eric89GXL) [inline mne/decoding/tests/test_csp.py]: Due to PCA/SVD equivalence I think this is equivalent to `mixing_mat = linalg.svd(rs.randn(n_channels, n_channels))[0]` if you want to save a line.

**Comment by alexandrebarachant** [comment]
I added a test (for Spoc and CSP) with simulated data to check if we recover the patterns from a instantaneous linear mixing model.
If travis is happy, we are good to go.

**Comment by alexandrebarachant** [comment]
Okay, the example takes too much memory. I will see how i can decrease it.
Also, i noticed a stange behaviour of CSP with simulation, the predictive patterns is sorted in last position. 
I have to check if it's a regression or if somehow its what is supposed to happens with my simulation.

**Comment by Eric89GXL** [comment]
> Okay, the example takes too much memory. I will see how i can decrease it.

Yeah I'd try just using a few channels (3?), it might be enough.

**Comment by alexandrebarachant** [comment]
No, defenetly a bug. pyRiemann sort the component first on the same data.
Good you pushed for a functional test :)

**Comment by alexandrebarachant** [comment]
Okay, i found the bug.
This is somehow due to the trace normalization of the covariance estimation in CSP : https://github.com/mne-tools/mne-python/blob/master/mne/decoding/csp.py#L170
with the current code and my simulation, the best CSP patterns is sorted last ?? no idea why, so when we do this : 

```python
csp = CSP(1)
out = csp.fit_transform(X, y);
plt.plot(scale(out[:, 0]))
plt.plot(scale(y))
```

image

now if i remove the trace normalization, and execute the same code, we get the expected behavior

image

What do you think ? should we remove the trance normalization ? 
We had the same problem with spoc and decided not to go for it.

**Comment by alexandrebarachant** [comment]
I think i understand the problem.
The trace normalization rescale the matrix to have a unit trace. However, this operation is applied only during the `fit`. when we transform the data, we just do the variance, with no normalization.

In my simulation, most of the power of the matrix is carried by one pattern, if we normalize the covariance, that will make the difference between condition vanish for the main pattern, and it project it back the non-discriminant patterns. So when we sort the components, it thinks the bad patterns are discriminant.

In real life that does not happens, but i will really suggest to get rid of the trace normalization.

**Comment by Eric89GXL** [comment]
Sounds reasonable to me. But why is the trace normalization there in the first place, then? Is there a paper that suggests doing it or something?

An alternative approach would be to normalize to unit trace during both fit and transform steps. But it sounds like you're saying that this will not work (at least for your simulated data)?

**Comment by alexandrebarachant** [comment]
This is something they do in some papers.
I traced this back to the original CSP papers (koles et al., 1990) with the following justification : 
> The normalization of R [NDLR Covariance] with respect to the trace
was done to eliminate magnitude variations in the EEG
between individuals. 

it is generally abandoned in more recent works.

I don't think we should do it in the transform either (and i don't think it is possible). If we follows my simulation, then it will screw up interpretation of the patterns (since the output will be predictive, but the patterns will not correspond to the generative pattern).

However, if i remove it, one test checking the rought equivalence between the 'epoch' and 'concat' covariance estimation will fail : https://github.com/mne-tools/mne-python/blob/master/mne/decoding/tests/test_csp.py#L89

with a correlation of ~0.8 instead of 0.94 between the two method

**Comment by Eric89GXL** [comment]
It seems okay to me to adjust the test, assuming it's an expected result of not normalizing by the trace.

Maybe we should expose a parameter, and deprecate `norm_trace=None` as `True` for this release but `False` for the next.

@kingjr WDYT?

**Comment by alexandrebarachant** [comment]
@agramfort might also have some thought about the trace normalization

I would prefer not exposing a parameter, the user don't really care about that. It would be actually cleaner to trance normalize the epochs before the CSP if we want to do this right.

Also, I will have to remove the test completely, it seems to change the order of the patterns so the test has no purpose anymore.

Let me benchmark the performances on a different dataset and see if it has an impact on decoding performances.

**Comment by alexandrebarachant** [comment]
Okay, trace normalization seems to decrease performances a little (with some big difference for some subjects)

image

if i remove it, all goes back to normal

image

The patterns are actually similar, the main difference is due to their ordering (and therefore their selection)

File: mne/decoding/csp.py [hunk 23]
```diff
@@ -171,7 +171,7 @@ def fit(self, X, y):
                 weight = len(class_)
 
             # normalize by trace and stack
```
Comment (Eric89GXL) [inline mne/decoding/csp.py]: This comment probably needs to change, too :)

Maybe just mention that trace normalization can break some use cases so we omit it here

Comment (alexandrebarachant) [inline mne/decoding/csp.py]: done.

**Comment by kingjr** [comment]
> Maybe we should expose a parameter, and deprecate norm_trace=None as True for this release but False for the next. @kingjr WDYT?

+1

Great benchmark @alexandrebarachant !

**Comment by alexandrebarachant** [comment]
@kingjr If you have a minute, can you add the changes you requested (it's your branch :) ) I will not have time to work on this today.

**Comment by kingjr** [comment]
@alexandrebarachant I get rank deficiencies in the CSP test if I set `norm_trace` to False. I suspect it's because of the way you did your simulation.

Could you have a look? I added `XXX` everywhere we encounter an error.

**Comment by alexandrebarachant** [comment]
@kingjr I don't have any error (or warning) on my side. I don't think it should be rank deficient, since i generated noise sources as well.
Can you push your commit ?

**Comment by kingjr** [comment]
Ok, that was an error on my part, it should work now.

**Comment by alexandrebarachant** [comment]
@kingjr your merge made the commit history and the list of file changed a bit hard to follow, but we where done anyway ...

Are we good to go ? I don't think there is anything more to be done here.

**Comment by kingjr** [comment]
Yeah.. using a different branch names ('spoc', vs 'jr/spoc') confused me...

The changes I introduced are just in the last commit, the rest is rebasing.

**Comment by agramfort** [comment]
@kingjr you made a mess in the git history. The diff contains really unrelated stuff. Please fix this.

**Comment by kingjr** [comment]
@alexandrebarachant this double branch turned out to be a nightmare. I started from scratch. I think I got everything, but can you review it?

**Comment by alexandrebarachant** [comment]
There was a bunch of file related to the new dataset.

I still have my local branch clean, just missing your last commit ...

**Comment by alexandrebarachant** [comment]
okay, i will force my branch, there is too many missing things. 
added your modification so it should be fine

**Comment by alexandrebarachant** [comment]
Okay, I think we are good.

ready for a merge on my side.

**Comment by kingjr** [comment]
Great, I just reviewed it again and it LGTM

@Eric89GXL  or @agramfort merge if you're happy

**Comment by Eric89GXL** [comment]
Thanks @alexandrebarachant
